src/yaml/from_yaml.cr

```diff
@@ -1,132 +1,160 @@
-def Object.from_yaml(string_or_io) : self
-  YAML::PullParser.new(string_or_io) do |parser|
-    parser.read_stream do
-      parser.read_document do
-        new parser
-      end
-    end
+def Object.from_yaml(string_or_io : String | IO) : self
+  new(YAML::ParseContext.new, parse_yaml(string_or_io))
+end
+
+def Array.from_yaml(string_or_io : String | IO)
+  new(YAML::ParseContext.new, parse_yaml(string_or_io)) do |element|
+    yield element
   end
 end
 
-def Array.from_yaml(string_or_io)
-  YAML::PullParser.new(string_or_io) do |parser|
-    parser.read_stream do
-      parser.read_document do
-        new(parser) do |element|
-          yield element
-        end
-      end
-    end
+private def parse_yaml(string_or_io)
+  document = YAML::Nodes.parse(string_or_io)
+
+  # If the document is empty we simulate an empty scalar with
+  # plain style, that parses to Nil
+  document.nodes.first? || begin
+    scalar = YAML::Nodes::Scalar.new("")
+    scalar.style = YAML::ScalarStyle::PLAIN
+    scalar
   end
 end
 
-def Nil.new(pull : YAML::PullParser)
-  location = pull.location
-  value = pull.read_scalar
-  if value.empty?
-    nil
+private def parse_scalar(ctx, node, type : T.class) forall T
+  ctx.read_alias(node, T) do |obj|
+    return obj
+  end
+
+  if node.is_a?(YAML::Nodes::Scalar)
+    value = YAML::Schema::Core.parse_scalar(node)
+    if value.is_a?(T)
+      ctx.record_anchor(node, value)
+      value
+    else
+      node.raise "Expected #{T}, not #{node.value}"
+    end
   else
-    raise YAML::ParseException.new("Expected nil, not #{value}", *location)
+    node.raise "Expected #{T}, not #{node.class.name}"
   end
 end
 
-def Bool.new(pull : YAML::PullParser)
-  pull.read_scalar == "true"
+def Nil.new(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
+  parse_scalar(ctx, node, self)
+end
+
+def Bool.new(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
+  parse_scalar(ctx, node, self)
 end
 
 {% for type in %w(Int8 Int16 Int32 Int64 UInt8 UInt16 UInt32 UInt64) %}
-  def {{type.id}}.new(pull : YAML::PullParser)
-    location = pull.location
-    begin
-      {{type.id}}.new(pull.read_scalar)
-    rescue ex
-      raise YAML::ParseException.new(ex.message.not_nil!, *location)
-    end
+  def {{type.id}}.new(ctx : YAML::ParseContext, node : YAML::Nodes::Node)
+    {{type.id}}.new parse_scalar(ctx, node, Int64)
```

Comment:
@asterite I don't follow how/if `UInt64` is supported. The call to parse_scalar uses `Int64` to check if the parsed type matches, but UInt64 won't fit.

```crystal
require "yaml"

typeof(YAML::Schema::Core.parse_scalar(UInt64::MAX.to_s)) # => Bool | Float64 | Int64 | String | Time | Nil
YAML::Schema::Core.parse_scalar(UInt64::MAX.to_s).class   # => Float64
```

Note that

```
it_parses_scalar UInt64::MAX.to_s, UInt64::MAX
```

passes since `UInt64::MAX == UInt64::MAX.to_f`

Reply:
I think it's supported if you do `UInt64.from_yaml(...)`, but `parse_scalar` only returns one of the `Type` values, so just `Int64` is supported (I don't know if it overflows or it just returns a string in this case).

Reply:
It returns a Float64 now :-)